### ISSUE TITLE
[Typescript] Complex arrow function as return type

### ISSUE THREAD

**Issue by rchl** [sublimehq/Packages#3059]
## Code

```ts
/**
 * A Format Locale Object
 */
export interface FormatLocaleObject {
    /**
     * Returns a new format function for the given string specifier. The returned function
     * takes a number as the only argument, and returns a string representing the formatted number.
     *
     * @param specifier A Specifier string.
     * @throws Error on invalid format specifier.
     */
    format(specifier: string): (n: number | { valueOf(): number }) => string;

    /**
     * Returns a new format function for the given string specifier. The returned function
     * takes a number as the only argument, and returns a string representing the formatted number.
     * The returned function will convert values to the units of the appropriate SI prefix for the
     * specified numeric reference value before formatting in fixed point notation.
     *
     * @param specifier A Specifier string.
     * @param value The reference value to determine the appropriate SI prefix.
     * @throws Error on invalid format specifier.
     */
    formatPrefix(specifier: string, value: number): (n: number | { valueOf(): number }) => string;
}
```

## Screenshot

![Screenshot 2021-09-30 at 12 20 36](https://user-images.githubusercontent.com/153197/135437836-ffa2ec52-ed5f-419d-97d9-48b4d126cdc4.png)

**Comment by thetutlage** [comment]
Another instance where the TypeScript syntax gets confused is when a class method returns an array of a type that accepts a generic.

<img width="629" alt="CleanShot_2021-09-30_at_17 05 182x" src="https://user-images.githubusercontent.com/1706381/135450376-ecabb167-3aa4-4151-a59a-8566e96f8953.png">

If I remove the `[]` then all works fine

<img width="455" alt="CleanShot_2021-09-30_at_17 07 182x" src="https://user-images.githubusercontent.com/1706381/135450414-9416f14e-3b5a-4666-bf80-69e12c754bf2.png">

**Comment by Thom1729** [comment]
Thanks. The original issue should be fixed in #3062, and the array modifier issue in #3063.
